src/client/Client.h

```diff
@@ -1298,4 +1323,133 @@ class StandaloneClient : public Client
   void shutdown() override;
 };
 
+class destructive_lock_ref_t {
+public:
+  destructive_lock_ref_t(const destructive_lock_ref_t &) = delete;
+  destructive_lock_ref_t(const destructive_lock_ref_t &&) = delete;
+  destructive_lock_ref_t(class Client *c, Client::_state require, bool update_state=false) {
```

Comment:
I feel this could be simplified by just using a `std::shared_mutex`. The difficulty is preventing starvation for the unmount thread. We can keep this class and add an atomic flag like `DESTROYING` or similar to prevent further read locks?

Making this a generic in `src/common` would also be appropriate.

Reply:
> I feel this could be simplified by just using a `std::shared_mutex`. The difficulty is preventing starvation for the unmount thread. We can keep this class and add an atomic flag like `DESTROYING` or similar to prevent further read locks?
> 

At first I have tried to use the shared mutex, but there was one case I didn't find any good way to avoid:

With the shared mutex, it will be like:

for unmounting(), the writer:

```
unmounting() {
    atomic_set(client_state, STATE_UNMOUNTING);
    scope_write_lock(shared_lock);  // it will wait if there has any in-flight reader running   
    
    ...

    client_shared_lock.unlock();
    cond.wait(client_lock, fn()); // before the cond.wait() we must unlock the shared lock, and grab it after this manually.   
    client_shared_lock.lock();    

    ...
}
```

For the readers:
```
reader_fun() {
    scope_read_lock(client_shared_lock); 
    if (atomic_read(client_state) == STATE_UNMOUNTING)
        return;
    
    client_lock.lock();
     ...

    client_shared_lock.unlock();
    cond.wait(client_lock, fn()); // before the cond.wait() we must unlock the shared lock, and grab it after this manually.   
    client_shared_lock.lock();    

    ...
    client_lock.unlock();

    ...
}
```

The biggest problem is we must manually do the `client_shared_lock` unlock/lock in all the `cond.wait` cases once the  `client_shared_lock` is locked.

The current approach won't have this problem, and it will simpler and easier to use. What we need to do in readers is at the begining to add:

```
read_func() {
    destructive_lock_ref_t ulr(this, Client::STATE_MOUNTED);  // require state to be mounted
    if (!ulr.is_satisified())
        return;

    ...
}
```

For the writer, it just need to update the state and then wait:

```
 _unmount() {
    destructive_lock_ref_t ulr(this, Client::STATE_UNMOUNTING, true);
    if (!ulr.unmount_wait()) // will wait all the "readers" to finish
        return; // the client is already umounting or unmounted

     ...
}
```


> Making this a generic in `src/common` would also be appropriate.

I will try this.

Thanks